test/e2e/tests/setup_e2e_test.go

```diff
@@ -61,28 +63,49 @@ var _ = BeforeSuite(func() {
 
 	fileInstancesService = filev1beta1.NewProjectsLocationsInstancesService(fileService)
 
-	for _, zone := range zones {
-		nodeID := fmt.Sprintf("gcfs-csi-e2e-%s", zone)
-
-		if *runInProw {
-			*project, *serviceAccount = testutils.SetupProwConfig()
-		}
-
-		Expect(*project).ToNot(BeEmpty(), "Project should not be empty")
-		Expect(*serviceAccount).ToNot(BeEmpty(), "Service account should not be empty")
+	if *runInProw {
+		*project, *serviceAccount = testutils.SetupProwConfig()
+	}
 
-		i, err := remote.SetupInstance(*project, zone, nodeID, *serviceAccount, computeService)
-		Expect(err).To(BeNil())
+	Expect(*project).ToNot(BeEmpty(), "Project should not be empty")
+	Expect(*serviceAccount).ToNot(BeEmpty(), "Service account should not be empty")
 
-		testInstances = append(testInstances, i)
+	for _, zone := range zones {
+		go func(curZone string) {
+			defer GinkgoRecover()
+			nodeID := fmt.Sprintf("gcfs-csi-e2e-%s", curZone)
+
+			i, err := remote.SetupInstance(*project, curZone, nodeID, *serviceAccount, computeService)
+			Expect(err).To(BeNil())
+
+			// Create new driver and client
+			testContext, err := testutils.GCFSClientAndDriverSetup(i)
+			Expect(err).To(BeNil(), "Set up new Driver and Client failed with error")
+			tcc <- testContext
+		}(zone)
 	}
 
+	for i := 0; i < len(zones); i++ {
+		tc := <-tcc
+		testContexts = append(testContexts, tc)
+	}
 })
 
 var _ = AfterSuite(func() {
-	for _, i := range testInstances {
-		if *deleteInstances {
-			i.DeleteInstance()
-		}
+	for _, tc := range testContexts {
+		go func(curTC *remote.TestContext) {
+			defer GinkgoRecover()
+			err := remote.TeardownDriverAndClient(curTC)
+			Expect(err).To(BeNil(), "Teardown Driver and Client failed with error")
+			if *deleteInstances {
+				curTC.Instance.DeleteInstance()
+			}
+		}(tc)
```

Comment:
Don't you have to wait for the teardown to finish? main() at least won't wait for other goroutines before existing; I'm not sure if that's true in test or not.

Reply:
I double checked the documentation and cannot find anything related to test being special with goroutines. I've added a waitgroup to account for this problem.

I also double checked the create logic and found that the create call currently can get stuck waiting forever if there is an error in the setup - I've changed it a little so that it will no longer hang, PTAL there as well!